Improved README.md file for can-utils

alan707 opened the pull request:
Hi @marckleinebudde and @hartkopp !
Thanks so much for maintaining can-utils!  I wanted to contribute back by making the README file a bit more accessible with more information on how to get started.  I know not all the information I've added on the README file is complete, but it is a starting point.  Here are the main things that I would like to get your input on:

- Added MIT License file called LICENSE.md - if that is not the correct license, feel free to change the content of the file.
- Under authors, I've put both your names.  Please correct if I was mistaken or missing anyone else.

Any additional input is welcome and curious what you think of this README file proposal.  Thanks for all your hard work!

dogmatic69 commented on README.md:
`````diff
@@ -1,17 +1,126 @@
-originally initially created by Robert Bosch GmbH. SocketCAN is a set
-of open source CAN drivers and a networking stack contributed by
-Volkswagen Research to the Linux kernel.
+````can-utils```` is an extremely versatile collection of CAN utilities to sniff, transmit, record, generate, and replay CAN messages. ````can-utils```` is based on a set of open source CAN drivers and networking stack called SocketCAN, contributed by Volkswagen Research to the Linux Kernel.
`````
Any reason for five backticks? One should be enough. \`foobar\` -> `foobar`

hartkopp replied:
Thanks Alan!
This brings me to the point to add a license file to the repository %-)
But in fact the idea was to have a BSD/GPL dual license (what you can see when looking into e.g. cansend.c) to be able to use the (kernel) code on BSD/QNX too. Don't know if it makes sense to copy the license information from the existing file - or use a vanilla GPL(v2)?
At least can-utils became part of Debian/Ubuntu due to the fact that it is GPL.

Regarding the 'quick intro' you proposed below:
I would suggest to extend http://elinux.org/CAN_Bus for this and refer to it then.

At least it would be ok to give a short summary about the tools that can be found in can-utils, e.g. like in this SocketCAN intro PDF I presented at AGL recently: https://wiki.automotivelinux.org/_media/agl-distro/agl2017-socketcan-print.pdf

Regards,
Oliver

alan707 commented:
Hey Oliver,
Thanks for the quick reply.  I've replaced the text from the GNU GPL v3 (latest version) and replaced the link that was pointing to Wikipedia's CAN bus article to to the link pointing to http://elinux.org/CAN_Bus.

Regarding licenses, here is a good website that I reference: [https://choosealicense.com/](https://choosealicense.com/)

Hope that is enough to get this pull request merged - I'm sure there are many improvements contributors can continue to contribute to make can-utils even more widespread!  Thanks again for this amazing toolkit!

alan707 commented:
Oliver, I also added the logo I found on the linux-can main repo, since I think it looks cool 😉

hartkopp commented:
Hi Alan,
I picked up many of your ideas to improve the README.md.
Although I will fill the descriptions the next days.
See the commit description.
Many thanks,
Oliver

alan707 commented:
Hey Oliver,
Any reason why this PR was not merged instead of copying and pasting the code?  It would be nice to contribute together, especially since I can offer the perspective of someone new trying to use these utilities.  It might help others from making the same mistake, that is why I spent a lot of time on that pull request.  Now I get a big "denied" mark.

Maybe next time just merge the pull request and change on top of that, so at least the community can also add things.

alan707 commented:
I want to keep can-utils going strong,
 so maybe we can adopt some contribution guidelines like the ones I proposed on my PR?

Interesting quote:

> In my experience, most healthy projects end up using one of two models: Contributions are either merged quickly and then refactored by the project's core developers, or the core developers provide constructive criticism and ask contributors to get their work into an acceptable state prior to being merged.
There's also an important caveat: Sometimes a new contributor's work isn't suitable for inclusion, but inspires others to write a superior implementation of similar functionality, in which case the original contributor should at least be referenced in the commit message.
If the project's author is constantly rewriting trivial patches and refusing to merge others' work directly, it sounds like your time would be better spent contributing elsewhere.

[https://www.reddit.com/r/opensource/comments/31dnft/question_good_or_bad_open_source_etiquette_you/](https://www.reddit.com/r/opensource/comments/31dnft/question_good_or_bad_open_source_etiquette_you/)

hartkopp commented:
Hi Alan,

it was not my intention to generate any frustration on your side - and I definitely appreciate your work to make can-utils better. But the pull request still contained the currently unresolved license discussion, it still contained generic 'CAN how-to' (e.g. slcan explanation) and it referenced some png logo which was not hosted inside the linux-can organization.

As you might see from the commits in can-utils many pull requests for code(!) are pulled directly, when it fits into the requirements of functionality and code style. Some pull requests are pushed back when it breaks the rules and intentions, see here https://github.com/linux-can/can-utils/pull/29 . This was a hard discussion for me and I do this in my spare time.
You started to 'design' the visual front end of can-utils - which finally pushed me to improve the README and to give you a big attribution for your work. If you would have prepared smaller steps and if you would have reacted on my advise to remove the generic CAN how-to you probably won't be frustrated about the current situation and the effort you spent.
I'm not the "project's author is constantly rewriting trivial patches". It just had to fit. And the complete package for the pull request did not fit (see above).
I'm still searching for volunteers to write man pages for can-utils - but this is probably not that fancy.

Don't be frustrated about this specific pull request. Open Source development is cool and lives from people pushing things to a better state. You got that done and you have been attributed for it.

Oliver

alan707 commented:
Hey Oliver,
Thanks for the thorough explanation, it actually makes sense what you are saying. I was always under the impression that merging is quite harmless since the repo owner can make changes on top of what has been pull requested from a contributor.  In my experience merging pull requests stimulates more improvements from the community and more activity is always better because it shows an up-to-date repo.  However, I totally agree that I probably should've consulted before submitting a pull request for something as front facing as the README.md file of the project!

I'm glad that we started this conversation, though, and I'd love to continue contributing (in smaller steps) to the project.  There are some things we can do to improve the documentation and I want to make sure that we have a healthy conversation beforehand so you feel comfortable with the changes.

I think an important piece that needs to be added to the README file is a "Contribution Guideline" so people understand how best to contribute if they have an idea on how to make it better.

Thanks for your hard work!

Alan
